File: 2015/c/src/day2.c
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "lib.h"

static int line_to_dims(int dims[3], char *line)
{
    dims[0] = atoi(strtok(line, "x"));
    if (dims[0] == 0)
        goto err_cleanup;

    dims[1] = atoi(strtok(NULL, "x"));
    if (dims[1] == 0)
        goto err_cleanup;

    dims[2] = atoi(strtok(NULL, "x"));
    if (dims[2] == 0)
        goto err_cleanup;

    qsort(dims, 3, sizeof(int), cmp_int);
    return 0;

err_cleanup:
    errno = EINVAL;
    return -1;
}

static int part1(const char *input, size_t input_len)
{
    int paper = 0;

    char **lines;
    int line_count = input_to_lines(&lines, input, input_len);
    if (line_count < 0)
        return -1;

    for (int i = 0; i < line_count; i++)
    {
        int dims[3];
        if (line_to_dims(dims, lines[i]))
            goto err_cleanup;

        paper += 2 * dims[0] * dims[1] + 2 * dims[1] * dims[2] + 2 * dims[2] * dims[0] + dims[0] * dims[1];
    }
    goto cleanup;

err_cleanup:
    paper = -1;
cleanup:
    free_lines(lines, line_count);
    return paper;
}

static int part2(const char *input, size_t input_len)
{
    int ribbon = 0;

    char **lines;
    int line_count = input_to_lines(&lines, input, input_len);
    if (line_count < 0)
        return -1;

    for (int i = 0; i < line_count; i++)
    {
        int dims[3];
        if (line_to_dims(dims, lines[i]))
            goto err_cleanup;

        ribbon += 2 * dims[0] + 2 * dims[1] + dims[0] * dims[1] * dims[2];
    }
    goto cleanup;

err_cleanup:
    ribbon = -1;
cleanup:
    free_lines(lines, line_count);
    return ribbon;
}

int day2(const char *filename)
{
    printf("Day 2: I Was Told There Would Be No Math\n");

    char *input;
    int filesize = read_file_to_buffer(&input, filename);
    if (filesize < 0)
        return -1;

    int rval = part1(input, filesize);
    if (rval < 0)
        return rval;
    printf("\tPart 1: %d\n", rval);

    rval = part2(input, filesize);
    if (rval < 0)
        return rval;
    printf("\tPart 2: %d\n", rval);

    free(input);

    return 0;
}

File: 2015/rust/src/day2.rs
use std::error::Error;
use std::fs::File;
use std::io::prelude::*;
use std::time::Instant;

pub fn day2(input_file: &str) -> Result<(), Box<dyn Error>> {
    println!("Day 2: I Was Told There Would Be No Math");

    let mut f = File::open(input_file)?;
    let mut input = String::new();
    f.read_to_string(&mut input)?;

    let start = Instant::now();
    println!("\tPart 1: {} ({:?})", part1(&input)?, start.elapsed());
    let second = Instant::now();
    println!("\tPart 2: {} ({:?})", part2(&input)?, second.elapsed());
    println!("\t\t Completed in {:?}", start.elapsed());
    Ok(())
}

fn input_to_dims(input: &str) -> Result<Vec<Vec<i32>>, Box<dyn Error>> {
    let mut dims = input
        .lines()
        .map(|line| {
            line.split("x")
                .map(|d| d.parse::<i32>())
                .collect::<Result<Vec<i32>, _>>()
        })
        .collect::<Result<Vec<Vec<i32>>, _>>()?;
    dims.sort();
    Ok(dims)
}

fn part1(input: &str) -> Result<i32, Box<dyn Error>> {
    Ok(input_to_dims(input)?.iter().fold(0, |paper, dims| {
        paper
            + 2 * dims[0] * dims[1]
            + 2 * dims[1] * dims[2]
            + 2 * dims[2] * dims[0]
            + dims[0] * dims[1]
    }))
}

fn part2(input: &str) -> Result<i32, Box<dyn Error>> {
    Ok(input_to_dims(input)?.iter().fold(0, |ribbon, dims| {
        ribbon + 2 * dims[0] + 2 * dims[1] + dims[0] * dims[1] * dims[2]
    }))
}
